ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs

```diff
@@ -1,58 +1,299 @@
+
+{-# COMPLETE LedgerErrorByron, LedgerErrorShelley, LedgerErrorWrongEra #-}
+
+{-------------------------------------------------------------------------------
+  OtherEnvelopeError
+-------------------------------------------------------------------------------}
+
+-- | An error resulting from validating a 'CardanoHeader'.
+--
+-- Thanks to the pattern synonyms, you can treat this as a sum type with
+-- constructors 'OtherHeaderEnvelopeErrorByron',
+-- 'OtherHeaderEnvelopeErrorShelley', and 'OtherHeaderEnvelopeErrorWrongEra'.
+type CardanoOtherHeaderEnvelopeError sc = HardForkEnvelopeErr (CardanoEras sc)
+
+pattern OtherHeaderEnvelopeErrorByron
+  :: OtherHeaderEnvelopeError ByronBlock
+  -> CardanoOtherHeaderEnvelopeError sc
+pattern OtherHeaderEnvelopeErrorByron err =
+    HardForkEnvelopeErrFromEra
+      (OneEraEnvelopeErr (Z (SingleEraEnvelopeErr err)))
+
+pattern OtherHeaderEnvelopeErrorShelley
+  :: OtherHeaderEnvelopeError (ShelleyBlock sc)
+  -> CardanoOtherHeaderEnvelopeError sc
+pattern OtherHeaderEnvelopeErrorShelley err =
+    HardForkEnvelopeErrFromEra (OneEraEnvelopeErr (S (Z (SingleEraEnvelopeErr err))))
+
+pattern OtherHeaderEnvelopeErrorWrongEra
+  :: Text
+  -> Text
+  -> CardanoOtherHeaderEnvelopeError sc
+pattern OtherHeaderEnvelopeErrorWrongEra txEraName ledgerEraName <-
+    HardForkEnvelopeErrWrongEra (mismatchEraNames -> (txEraName, ledgerEraName))
+
+{-# COMPLETE OtherHeaderEnvelopeErrorByron
+           , OtherHeaderEnvelopeErrorShelley
+           , OtherHeaderEnvelopeErrorWrongEra #-}
+
+{-------------------------------------------------------------------------------
+  TipInfo
+-------------------------------------------------------------------------------}
+
+-- | The 'TipInfo' of Cardano chain.
+--
+-- Thanks to the pattern synonyms, you can treat this as a sum type with
+-- constructors 'TipInfoByron' and 'TipInfoShelley'.
+type CardanoTipInfo sc = OneEraTipInfo (CardanoEras sc)
+
+pattern TipInfoByron :: TipInfo ByronBlock -> CardanoTipInfo sc
+pattern TipInfoByron ti = OneEraTipInfo (Z (SingleEraTipInfo ti))
+
+pattern TipInfoShelley :: TipInfo (ShelleyBlock sc) -> CardanoTipInfo sc
+pattern TipInfoShelley ti = OneEraTipInfo (S (Z (SingleEraTipInfo ti)))
+
+{-# COMPLETE TipInfoByron, TipInfoShelley #-}
+
+{-------------------------------------------------------------------------------
+  CodecConfig
+-------------------------------------------------------------------------------}
+
+-- | The 'CodecConfig' for 'CardanoBlock'.
+--
+-- Thanks to the pattern synonyms, you can treat this as the product of
+-- the Byron and Shelley 'ConsensusConfig'.
+type CardanoCodecConfig sc = CodecConfig (CardanoBlock sc)
+
+pattern CardanoCodecConfig
+  :: CodecConfig ByronBlock
+  -> CodecConfig (ShelleyBlock sc)
+  -> CardanoCodecConfig sc
+pattern CardanoCodecConfig cfgByron cfgShelley =
+    HardForkCodecConfig (PerEraCodecConfig (cfgByron :* cfgShelley :* Nil))
+
```

Comment:
Add `CardanoTopLevelConfig`

Create function for the `protocolInfo` and `protocolClientInfo` for `CardanoBlock`. Not necessary for the node-facing API, but just for us, as we have to give the `protocolInfo` and `protocolClientInfo`.

Reply:
We can't easily do `CardanoTopLevelConfig`, because it contains partial configs.